### ISSUE TITLE
Nagios RESTART_PROGRAM external command defunct processes

### ISSUE THREAD

**Issue by douglasawh** [NagiosEnterprises/nagioscore#441]
Using the script provided at https://old.nagios.org/developerinfo/externalcommands/commandinfo.php?command_id=9 the issue reported in https://support.nagios.com/forum/viewtopic.php?t=46081&start=9999#bottom is reproducible on a system with a single host, and the default services following the Core/Plugins instructions at https://support.nagios.com/kb/article.php?id=96 for Fedora.

`ps -aef | grep nagios` output after restarting with init:
> nagios   11211     1  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios -d /usr/local/nagios/etc/nagios.cfg
> nagios   11214 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11215 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11216 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11217 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11218 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11219 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11227 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios -d /usr/local/nagios/etc/nagios.cfg
> root     11229  2089  0 12:53 pts/0    00:00:00 grep --color=auto nagios

`ps -aef | grep nagios` output after external program:
> nagios   11211     1  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios -d /usr/local/nagios/etc/nagios.cfg
> nagios   11214 11211  0 12:53 ?        00:00:00 [nagios] <defunct>
> nagios   11215 11211  0 12:53 ?        00:00:00 [nagios] <defunct>
> nagios   11216 11211  0 12:53 ?        00:00:00 [nagios] <defunct>
> nagios   11217 11211  0 12:53 ?        00:00:00 [nagios] <defunct>
> nagios   11218 11211  0 12:53 ?        00:00:00 [nagios] <defunct>
> nagios   11219 11211  0 12:53 ?        00:00:00 [nagios] <defunct>
> nagios   11227 11211  0 12:53 ?        00:00:00 /usr/local/nagios/bin/nagios -d /usr/local/nagios/etc/nagios.cfg
> nagios   11739 11211  0 14:06 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11740 11211  0 14:06 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11741 11211  0 14:06 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11742 11211  0 14:06 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11743 11211  0 14:06 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> nagios   11744 11211  0 14:06 ?        00:00:00 /usr/local/nagios/bin/nagios --worker /usr/local/nagios/var/rw/nagios.qh
> root     11748  2089  0 14:06 pts/0    00:00:00 grep --color=auto nagios

**Comment by rhassing** [comment]
I can reproduce it on CentOS 7 and Rasbian as well

**Comment by rhassing** [comment]
No news?

**Comment by hedenface** [comment]
Haven't tested it yet. Apologies, but there is quite a bit of backlog. Working on #46 and and #23 at the moment. I'll add the 4.4 milestone so I at least get to it before that is released.

**Comment by hedenface** [comment]
Confirmed. On linux mint even. Will try and get the fix out shortly.

**Comment by hedenface** [comment]
The bad part is that although an init restart works fine, the reload does the exact same thing. For reference, both the ext command and the reload command send a `SIGHUP` to Nagios, and once that happens the regrestart takes over.

**Comment by hedenface** [comment]
Fixed in https://github.com/NagiosEnterprises/nagioscore/commit/06e66eb29f0ebfecd287b222fcc89537c8972e18

**Comment by dvoryanchikov** [comment]
Hi!
I had this problem too, after first reload all workers are going to Z state until parent program stops/restarts.
Had this bug in 4.3.2 and 4.4.2 versions.

**Workaround**
I tried revert changes in `base/utils.c` from commit https://github.com/NagiosEnterprises/nagioscore/commit/79b239480458730bcb9c9e1e56275fe2169692c0#diff-8f96e94455e46eacb3ec583fcba0c435 
in 4.4.2 code and this fixed problem for me, no more zombies after program reload

```
--- a/base/utils.c	2018-08-24 01:30:00.000000000 +0300
+++ b/base/utils.c	2018-08-27 23:57:31.000000000 +0300
@@ -1637,6 +1637,7 @@
 	if(daemon_dumps_core == FALSE && daemon_mode == TRUE)
 		sigaction(SIGSEGV, &sig_action, NULL);
 	sig_action.sa_flags = SA_NOCLDWAIT;
+	sigaction(SIGCHLD, &sig_action, NULL);
 #else /* HAVE_SIGACTION */
 	signal(SIGPIPE, SIG_IGN);
 	signal(SIGQUIT, sighandler);
@@ -1644,6 +1645,7 @@
 	signal(SIGHUP, sighandler);
 	if(daemon_dumps_core == FALSE && daemon_mode == TRUE)
 		signal(SIGSEGV, sighandler);
+	signal(SIGCHLD, sighandler);
 #endif /* HAVE_SIGACTION */
 
 	return;
@@ -1703,6 +1705,18 @@
 	if(sig == SIGHUP)
 		sigrestart = TRUE;
 
+	else if(sig == SIGCHLD) {
+
+		pid_t child_pid;
+		int status = 0;
+
+		log_debug_info(DEBUGL_PROCESS, 0, "Caught SIGCHLD, calling waitpid() with WNOHANG\n");
+
+		while ((child_pid = waitpid(-1, &status, WNOHANG)) > 0) {
+			log_debug_info(DEBUGL_PROCESS, 0, " * waitpid() on child_pid = (%d)\n", (int)child_pid);
+		}
+	}
+
 	/* else begin shutting down... */
 	else if(sig < 16) {
 		logit(NSLOG_PROCESS_INFO, TRUE, "Caught SIG%s, shutting down...\n", sigs[sig]);
```
@hedenface why did u remove the sigchld signal handlers?

__
BR,
Alexey

**Comment by hedenface** [comment]
@dvoryanchikov  Those were only added recently. And they were causing problems.

They were not present in 4.3.1. https://github.com/NagiosEnterprises/nagioscore/blob/ca3ddc20b2cce7ca97a34fbb67b59ab0b0c93716/base/utils.c#L1664-L1694

They did in fact fix a zombie issue, but they caused another issue on some systems. I can't find it right this second, but when I do I will update the thread.

**Comment by gmourani** [comment]
Bryan,

So better to wait for an official fix on this issue ?

Gerhard,

**Comment by hedenface** [comment]
As a very simple workaround, you could just an actual restart instead of a reload.

But yes, I will investigate and solve. The sigchld with a bunch of `waitpid()s` was a quick solution I came up with to solve to the original problem - but it definitely had its drawbacks.

**Comment by pradeepspa2** [comment]
I am having the same issue with NagiosQL. Any fix or workaround?

**Comment by borisv1979** [comment]
I'm having this issue using Nagios Core 4.4.2 running on CentOS 6.10. It should be resolved in version 4.4.0, but it looks like problem still remains ? Any suggestions to fix this ?

**Comment by nikband** [comment]
I have the same problem with 4.4.3 and CentOS 7.4 
nagios    82057  0.0  0.0  10796  1016 ?        S    17:21   0:00 /nagios/bin/nagios --worker /nagios/var/rw/nagios.qh
nagios   182152  1.2  0.2 955304 74160 ?        Ssl  Apr16  14:08 /nagios/bin/nagios -d /nagios/etc/nagios.cfg
nagios   182188  0.0  0.0      0     0 ?        Z    Apr16   0:42 [nagios] <defunct>
nagios   182200  0.0  0.0 152228 10108 ?        S    Apr16   0:03 /nagios/bin/nagios -d /nagios/etc/nagios.cfg

**Comment by Madlohe** [comment]
Just wanted to offer an update on this issue. I was able to "solve" this by putting a `sleep(1);` just ahead of the zombie cleanup code. That block (~900-918 in nagios.c) now looks like:
```
	/* try and collect any zombie processes */
	if (sigrestart == TRUE) {

		sleep(1);

		int status = 0;
		pid_t child_pid;
		log_debug_info(DEBUGL_PROCESS, 1, "Calling waitpid() on all children...\n");

		while ((child_pid = waitpid(-1, &status, WNOHANG)) > 0) {

			log_debug_info(DEBUGL_PROCESS, 2, " * child PID: (%d), status: (%d)\n", child_pid, status);
			}

		log_debug_info(DEBUGL_PROCESS, 1, "All children have been wait()ed on\n");

		cleanup();
		}
```

This probably isn't the "right" fix, but we can probably put it in if I don't find a better one soon. Either way, this should be fixed for 4.4.4
